src/cose/mac0.rs

```diff
@@ -0,0 +1,410 @@
+        let tag_payload = mac_structure_data(
+            MacContext::CoseMac0,
+            cose_mac0.protected.clone(),
+            aad.unwrap_or_default(),
+            &payload,
+        );
+
+        Ok(Self {
+            cose_mac0: MaybeTagged::new(tagged, cose_mac0),
+            tag_payload,
+        })
+    }
+
+    /// Returns the signature payload that needs to be used to tag it.
+    pub fn signature_payload(&self) -> &[u8] {
+        &self.tag_payload
+    }
+
+    /// Finalize the COSE_Mac0 by adding the tag.
+    pub fn finalize(self, tag: Vec<u8>) -> MaybeTagged<CoseMac0> {
+        let mut cose_mac0 = self.cose_mac0;
+        cose_mac0.inner.tag = tag;
+        cose_mac0
+    }
+}
+
+impl MaybeTagged<CoseMac0> {
+    /// Verify that the tag of a `COSE_Mac0` is authentic.
+    pub fn verify(
+        &self,
+        verifier: &Hmac<Sha256>,
+        detached_payload: Option<&[u8]>,
+        external_aad: Option<&[u8]>,
+    ) -> VerificationResult {
+        if let Some(RegisteredLabelWithPrivate::Assigned(alg)) =
+            self.inner.protected.header.alg.as_ref()
+        {
+            if verifier.algorithm() != *alg {
+                return VerificationResult::Failure(
+                    "algorithm in protected headers did not match verifier's algorithm".into(),
+                );
+            }
+        }
+
+        let payload = match (self.inner.payload.as_ref(), detached_payload) {
+            (None, None) => return VerificationResult::Error(Error::NoPayload),
+            (Some(attached), None) => attached,
+            (None, Some(detached)) => detached,
+            _ => return VerificationResult::Error(Error::DoublePayload),
+        };
+
+        let tag = &self.inner.tag;
+
+        // Create the signature payload ot be used later on signing.
+        let tag_payload = mac_structure_data(
+            MacContext::CoseMac0,
+            self.inner.protected.clone(),
+            external_aad.unwrap_or_default(),
+            payload,
+        );
+
+        let mut mac = verifier.clone();
+        mac.reset();
+        mac.update(&tag_payload);
+        match mac.verify_slice(tag) {
+            Ok(()) => VerificationResult::Success,
+            Err(e) => VerificationResult::Failure(format!("tag is not authentic: {}", e)),
+        }
+    }
+
+    /// Retrieve the CWT claims set.
+    pub fn claims_set(&self) -> Result<Option<ClaimsSet>> {
+        match self.inner.payload.as_ref() {
+            None => Ok(None),
+            Some(payload) => ClaimsSet::from_slice(payload).map_or_else(
+                |e| Err(Error::UnableToDeserializeIntoClaimsSet(e)),
+                |c| Ok(Some(c)),
+            ),
+        }
+    }
+
+    /// If we are serialized as tagged.
+    pub fn is_tagged(&self) -> bool {
+        self.tagged
+    }
+
+    /// Set serialization to tagged.
+    pub fn set_tagged(&mut self) {
+        self.tagged = true;
+    }
```

Comment:
These could be methods on `MaybeTagged<T>` instead of `MaybeTagged<CoseMac0>`

Reply:
moved